Firestore/core/src/firebase/firestore/remote/grpc_stream.cc

```diff
@@ -0,0 +1,337 @@
+// `GrpcStreamDelegate`). The stream is guaranteed to be valid as long as the
+// operation exists.
+class StreamOperation : public GrpcOperation {
+ public:
+  StreamOperation(GrpcStreamDelegate&& delegate,
+                  grpc::GenericClientAsyncReaderWriter* call,
+                  GrpcCompletionQueue* grpc_queue)
+      : delegate_{std::move(delegate)}, call_{call}, grpc_queue_{grpc_queue} {
+  }
+
+  void Execute() override {
+    if (!grpc_queue_->IsShutDown()) {
+      DoExecute(call_);
+    }
+  }
+
+  void Complete(bool ok) override {
+    if (ok) {
+      DoComplete(&delegate_);
+    } else {
+      // Failed operation means this stream is irrecoverably broken; use the
+      // same error-handling policy for all operations.
+      delegate_.OnOperationFailed();
+    }
+  }
+
+ private:
+  virtual void DoExecute(grpc::GenericClientAsyncReaderWriter* call) = 0;
+  virtual void DoComplete(GrpcStreamDelegate* delegate) = 0;
+
+  // Delegate contains a strong reference to the stream.
+  GrpcStreamDelegate delegate_;
+  grpc::GenericClientAsyncReaderWriter* call_ = nullptr;
+  // Make execution a no-op if the queue is shutting down.
+  GrpcCompletionQueue* grpc_queue_ = nullptr;
+};
+
+class StreamStart : public StreamOperation {
+ public:
+  using StreamOperation::StreamOperation;
+
+ private:
+  void DoExecute(grpc::GenericClientAsyncReaderWriter* call) override {
+    call->StartCall(this);
+  }
+
+  void DoComplete(GrpcStreamDelegate* delegate) override {
+    delegate->OnStart();
+  }
+};
+
+class StreamRead : public StreamOperation {
+ public:
+  using StreamOperation::StreamOperation;
+
+ private:
+  void DoExecute(grpc::GenericClientAsyncReaderWriter* call) override {
+    call->Read(&message_, this);
+  }
+
+  void DoComplete(GrpcStreamDelegate* delegate) override {
+    delegate->OnRead(message_);
+  }
+
+  grpc::ByteBuffer message_;
+};
+
+class StreamWrite : public StreamOperation {
+ public:
+  StreamWrite(GrpcStreamDelegate&& delegate,
+              grpc::GenericClientAsyncReaderWriter* call,
+              GrpcCompletionQueue* grpc_queue,
+              grpc::ByteBuffer&& message)
+      : StreamOperation{std::move(delegate), call, grpc_queue},
+        message_{std::move(message)} {
+  }
+
+ private:
+  void DoExecute(grpc::GenericClientAsyncReaderWriter* call) override {
+    call->Write(message_, this);
+  }
+
+  void DoComplete(GrpcStreamDelegate* delegate) override {
+    delegate->OnWrite();
+  }
+
+  // Note that even though `grpc::GenericClientAsyncReaderWriter::Write` takes
+  // the byte buffer by const reference, it expects the buffer's lifetime to
+  // extend beyond `Write` (the buffer must be valid until the completion queue
+  // returns the tag associated with the write, see
+  // https://github.com/grpc/grpc/issues/13019#issuecomment-336932929, #5).
+  grpc::ByteBuffer message_;
+};
+
+class ServerInitiatedFinish : public StreamOperation {
```

Comment:
This name is a little misleading, because it might also be triggered by loss of connection/etc. What I really want is the opposite of "client initiated".